modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb

```diff
@@ -61,7 +61,11 @@ def legacy_classes_format(response)
 
   def get_classes(environment)
     future = async_get_classes(environment)
-    future.value!(MAX_CLIENT_TIMEOUT)
+    cache_used = @m.synchronize { !!@etag_cache[environment] } #etags are only available when classes cache is enabled
+    client_timeout = cache_used ? DEFAULT_CLIENT_TIMEOUT : 2 * DEFAULT_CLIENT_TIMEOUT
```

Comment:
I think the client timeout if no cache is used should be equal to the API timeout (300 seconds), rather than 30 seconds. This would be closer to the behaviour of the resource_types provider, and ensures the client gets the response. (If no caching's in use, the response will always be lost.)